SimulationAgent/Agent.cs

```diff
@@ -9,271 +10,426 @@
+                    this.devicesTelemetryTasks.Add(task);
+
+                    this.devicesTelemetryThreads[i] = new Thread(
+                        () => task.RunAsync(this.deviceTelemetryActors, i + 1, count, this.runningToken.Token));
+                    this.devicesTelemetryThreads[i].Start();
                 }
             }
+            catch (Exception e)
+            {
+                this.log.Error("Unable to start the device-telemetry threads", e);
+                throw new Exception("Unable to start the device-telemetry threads", e);
+            }
         }
 
-        private async Task CheckForStopOrStartToSimulationAsync()
+        private void TryToStopThreads()
         {
-            // stopped
-            if (this.simulation != null && this.simulation.Enabled && !this.simulation.ShouldBeRunning)
+            // State
+            try
+            {
+                this.devicesStateThread.Interrupt();
+            }
+            catch (Exception e)
             {
-                this.simulation.Statistics.AverageMessagesPerSecond = this.rateReporter.GetThroughputForMessages();
-                this.simulation.Statistics.TotalMessagesSent = this.runner.TotalMessagesCount;
+                this.log.Warn("Unable to stop the devices state thread in a clean way", e);
+            }
 
-                this.runner.Stop();
+            // Connection
+            try
+            {
+                this.devicesConnectionThread.Interrupt();
+            }
+            catch (Exception e)
+            {
+                this.log.Warn("Unable to stop the connections thread in a clean way", e);
+            }
 
-                // Update simulation statistics in storage
-                await this.simulations.UpsertAsync(this.simulation);
+            // Properties
+            try
+            {
+                this.devicesPropertiesThread.Interrupt();
+            }
+            catch (Exception e)
+            {
+                this.log.Warn("Unable to stop the devices state thread in a clean way", e);
             }
 
-            // started
-            if (this.simulation != null && this.simulation.ShouldBeRunning)
+            // Telemetry
+            this.devicesTelemetryTasks.Clear();
+            for (int i = 0; i < this.devicesTelemetryThreads.Length; i++)
             {
-                await this.runner.StartAsync(this.simulation);
+                try
+                {
+                    if (this.devicesTelemetryThreads[i] != null)
+                    {
+                        this.devicesTelemetryThreads[i].Interrupt();
+                    }
+                }
+                catch (Exception e)
+                {
+                    this.log.Warn("Unable to stop the telemetry thread in a clean way", () => new { threadNumber = i, e });
+                }
             }
         }
 
-        private bool IsDeviceModelIdValidAsync(string modelId)
+        private void SendSolutionHeartbeatAsync()
```

Comment:
please remove the "Async" suffix here. Internally the LogServiceHeartbeat is async, but the detail is not exposed

Reply:
Fixed.